Improve parsing of entropy (and enthalpy and freeenergy)

schneiderfelipe opened the pull request:
This should fix #663.

schneiderfelipe commented on test/data/testvib.py:
```diff
@@ -130,7 +130,7 @@ def testenthalpy(self):
     def testentropy(self):
          """Is the entropy reasonable"""
-         self.assertAlmostEqual(0.03601749, self.data.entropy, self.entropy_places)
+         self.assertAlmostEqual(0.00012080325339594164, self.data.entropy, self.entropy_places)
```
Aren't we rounding too soon with `entropy_places = 3` (line 116)?

schneiderfelipe replied:
There are uneven rounding places for entropy across programs. Gamess for instance is tested against 6 decimal places.

langner replied:
What do you recommend we do instead?

schneiderfelipe replied:
As long as the numbers are reliable, we could attempt to be as strict as possible: maximum number of common digits. (The figures I came up with were calculated independently so I believe they are reliable.)

Another possibility would be to set an energy/entropy cut, e.g. 1e-3 hartree (1e-4 hartree) ensures an error smaller than 0.63 kcal/mol (0.063 kcal/mol).

Entropy cuts need then to be calculated separately: by using a cut of 1e-3 hartree/K, we are accepting errors around 187 kcal/mol at 298.15 K! We need 1e-6 hartree/K (1e-7 hartree/K) for errors less than 0.2 kcal/mol (0.02 kcal/mol).

 I prefer 1e-4 hartree and 1e-7 hartree/K.

langner replied:
OK then, do you want to make the default for `entropy_places` more stringent and loosen it for select packages?

schneiderfelipe replied:
Yes, just to be sure that the parsing is working within a precise but reasonable energy deviation.

langner commented on test/data/testvib.py:
```diff
@@ -130,7 +130,7 @@ def testenthalpy(self):
     def testentropy(self):
          """Is the entropy reasonable"""
-         self.assertAlmostEqual(0.03601749, self.data.entropy, self.entropy_places)
+         self.assertAlmostEqual(0.00012080325339594164, self.data.entropy, self.entropy_places)
```
Do you know what the units here are? It is kcal/(mol K) like the logfile suggests?

schneiderfelipe replied:
True. I insisted a bit in hartree/(particle*kelvin) to keep consistency with other fields in the docs such as `enthalpy` and `freeenergy`, but we can change them all. Essentially, "particle" here means "unity".

langner replied:
SGTM

langner commented on test/data/testvib.py:
```diff
@@ -67,6 +67,16 @@ def testvibsyms(self):
+
+    def testentropy(self):
+         """Is the entropy reasonable"""
+         self.assertAlmostEqual(0.0001462623335480945, self.data.entropy, self.entropy_places)
```
This looks like a test we could move to GenericIRTest, and just customize the value in the package-specific test classes.

schneiderfelipe replied:
Which tests do you think we could move to GenericIRTest? We could move tests for temperature and pressure as well, for instance.

langner replied:
I was thinking only about the entropy at this time, since that is what you are changing.

langner commented on cclib/parser/data.py:
```diff
@@ -35,7 +35,7 @@ class ccData(object):
         coreelectrons -- number of core electrons in atom pseudopotentials (array[1])
         enthalpy -- sum of electronic and thermal enthalpies (float, hartree/particle)
-        entropy -- entropy (float, hartree/particle)
+        entropy -- entropy (float, hartree/particle*kelvin)
```
1. Is it customary to add the "particle" bit for entropy?
2. Shouldn't his be hartree/(particle*Kelvin) - currently looks a little like K is in the numerator if read programmatically.

schneiderfelipe replied:
>     1. Is it customary to add the "particle" bit for entropy?

I used to keep consistency with [the docs](https://cclib.github.io/data.html). I think it is fine if we remove.

>     2. Shouldn't his be hartree/(particle*Kelvin) - currently looks a little like K is in the numerator if read programmatically.

You're right. I'll fix that.

langner replied:
I don't mind. Actually I would defer to your judgement whether the particle bit should stay, just would press on trying to keep being consistent throughout the docs and any comments (it's always a struggle).

schneiderfelipe replied:
OK. So I would say to just add the parenthesis and leave the "particle" as it is.

langner replied:
Sounds good.

langner commented on cclib/parser/molcasparser.py:
```diff
@@ -483,7 +483,8 @@ def extract(self, inputfile, line):
                         line = next(inputfile)
-                    entropy_values.append(utils.convertor(float(line.split()[2]), 'kcal/mol', 'hartree'))
+                    # Molcas reports entropy values in kcal/mol*K but actually writes them in cal/mol*K
+                    entropy_values.append(utils.convertor(float(line.split()[2]), 'kcal/mol', 'hartree') / 1000)
```
Do we need to sum the entropy components for Moclas as well?

schneiderfelipe replied:
I don't think so. The output of Molcas produces values for a set of temperatures, we end up getting the entropy value for 298.15 K in https://github.com/cclib/cclib/blob/b50fdfd557077e9dfbd3d1605c07b5f82d90fb04/cclib/parser/molcasparser.py#L495-L513

Is that what you mean?

langner replied:
I see. And is there something we can do if 298.15 is not there but other temperature are?

schneiderfelipe replied:
Not sure, I've never used Molcas. Is it possible that 298.15 K is always reported?

langner replied:
I guess the comment says it all. Looks fine as far as your changes are concerned.

langner commented:
@schneiderfelipe sorry but there seem to have been conflicts merged in recently - possibly mine.

You need to merge or rebase manually - let me know if you need help with that.

schneiderfelipe commented:
This is failing and I don't know why. Errors are not due to decimal places as they didn't change after the last commit.

Failures are due to `dvb_ir.out` ORCA logfiles, but the newest version (4.0) is the only one that doesn't fail. In summary, errors happen with (2 failed tests each):

- `ORCA2.6/dvb_ir.out`, associated with class `OrcaIRTest_old_coordsOK` (inherits from `OrcaIRTest`)
- `ORCA2.8/dvb_ir.out`, associated with class `OrcaIRTest_old` (inherits from `OrcaIRTest` too)
- `ORCA2.9/dvb_ir.out`, associated with class `OrcaIRTest`
- `ORCA3.0/dvb_ir.out`, associated with class `OrcaIRTest` as well

`ORCA4.0/dvb_ir.out` presents no error, but is **associated with class `GenericIRTest`**, which doesn't test for entropy/enthalpy/freeenergy values. Furthermore, `OrcaIRTest` inherits from `GenericIRTest`.

This is clearly inconsistent. Is there a reason for this? What should we do?

langner commented:
> This is failing and I don't know why. Errors are not due to decimal places as they didn't change after the last commit.
> 
> Failures are due to `dvb_ir.out` ORCA logfiles, but the newest version (4.0) is the only one that doesn't fail. In summary, errors happen with (2 failed tests each):
> 
> * `ORCA2.6/dvb_ir.out`, associated with class `OrcaIRTest_old_coordsOK` (inherits from `OrcaIRTest`)
> * `ORCA2.8/dvb_ir.out`, associated with class `OrcaIRTest_old` (inherits from `OrcaIRTest` too)
> * `ORCA2.9/dvb_ir.out`, associated with class `OrcaIRTest`
> * `ORCA3.0/dvb_ir.out`, associated with class `OrcaIRTest` as well
> 
> `ORCA4.0/dvb_ir.out` presents no error, but is **associated with class `GenericIRTest`**, which doesn't test for entropy/enthalpy/freeenergy values. Furthermore, `OrcaIRTest` inherits from `GenericIRTest`.
> 
> This is clearly inconsistent. Is there a reason for this? What should we do?

Well, I think the cause of this is the improved accuracy in your checks. We've changed the coordinates and such over the years so that unit test logfiles for different packages are more aligned. So I'm not surprised we would see slight differences for these older logfiles which are now in the regression suite.

You can adjust the test class attributes in a custom test class and assign it to the old unit test in the regression suite. See this for an example: https://github.com/cclib/cclib/blob/master/test/regression.py#L2927. I know this is a little clumsy, and we need to improve the usability, plus that file is getting mighty large... and that will eventually get improved.

P. S. Trying to add more traceback information to the regression suite run on Travis in #834. Also, for some reason I don't see the Travis links/icons in github PRs - do you guys see them?

langner commented:
> > Well, I think the cause of this is the improved accuracy in your checks.
> 
> I don't think so. This is the [Travis build](https://travis-ci.org/github/cclib/cclib/builds/671655276) for [1ed29be](https://github.com/cclib/cclib/commit/1ed29bee41d2b00a5fb3fe220bb5ab7b7baf8ded) (prior to the change in accuracy). Here is the [Travis build for the last commit](https://travis-ci.org/github/cclib/cclib/builds/671669604) ([96e33a2](https://github.com/cclib/cclib/commit/96e33a22b291d82e79c7e95025f85637b18b31fa)), including the improved accuracy. The errors are the same.

Hmm.. need to look into this then.

> > P. S. Trying to add more traceback information to the regression suite run on Travis in #834.
> 
> Thanks. Let's see what the traceback info gives us.

Unfortunately my PR broke Travis so I reverted it. Need to figure out how to pass an additional flag using pytest.

> > Also, for some reason I don't see the Travis links/icons in github PRs - do you guys see them?
> 
> I don't see anything as well.

Seems there was some kind of migration from travis-ci.org to travis-ci.com. Removed and re-added the app integration and works again now from the new domain.

schneiderfelipe commented:
> Unfortunately my PR broke Travis so I reverted it. Need to figure out how to pass an additional flag using pytest.

I think adding something like

```
[pytest]
addopts = --tb=long
```

to [`pytest.ini` might work](https://docs.pytest.org/en/latest/customize.html#adding-default-options).

> Seems there was some kind of migration from travis-ci.org to travis-ci.com. Removed and re-added the app integration and works again now from the new domain.

Yes, I see it now. Thanks!

langner commented:
> > Unfortunately my PR broke Travis so I reverted it. Need to figure out how to pass an additional flag using pytest.
> 
> I think adding something like
> 
> ```
> [pytest]
> addopts = --tb=long
> ```

Unfortunately not. I'm trying to add use a flag that is defined in `regression.py`.

schneiderfelipe commented:
> Unfortunately not. I'm trying to add use a flag that is defined in `regression.py`.

Sorry, I thought that was the pytest `--tb` flag!

berquist reviewed (approved):
Once this is rebased and the CI failure is fixed, I'm happy with this.

schneiderfelipe commented on cclib/parser/orcaparser.py:
```diff
@@ -947,7 +947,7 @@ def splitter(line):
             # ORCA prints -inf for single atom entropy.
             if self.natom > 1:
-                self.entropy = float(next(inputfile).split()[4])
+                self.entropy = float(next(inputfile).split()[4]) / self.temperature
```
ORCA reports entropy as `T S`.

berquist reviewed (approved):
Sorry, I was hacking on this to remove most of the duplication in the inherited tests, then I think I discovered some other issues, not sure...but that's not worth dragging out this particular PR any further. Thanks for doing this!
